docs/variants_from_haplotypes.tex

```diff
@@ -0,0 +1,66 @@
+
+As in our pair-HMM probabilistic alignment, we fill a matrix $M$, the rows and columns of which correspond to bases of the reference and alternate haplotypes, from top to bottom and left to right.  $M_{ij}$ is the best score of alignments ending at the $i$th reference base and $j$th alternate base that do not end in a gap.  We also keep track of two arrays pertaining to the last row of our traversal.  $D_j$ is the best score of alignments ending at the previous reference base (ie the $(i-1)$th base when we are at the $i$th base in traversal) and the $j$th alternate base that end in a ``downward'' gap, i.e. a deletion with respect to the reference.  $S^d_j$ is the size of the gap in this best-scoring alignment.  We also fill a backtrack matrix $B$, where $B_{ij}$ is an instruction (see below) for reconstructing the best path after we fill $M$.
+
+First we initialize the zeroth row and column as $M_{0,0}$ = 0, $M_{0,1} = M_{1,0} = w_{\rm open}$, $M_{0,2} = M_{2,0} = w_{\rm open} + w{\rm extend}$ etc.  The zeroth row and column correspond to one base before the reference and alternate starts and this initialization penalizes leading indels.  That is, this is a global alignment\footnote{The GATK assembly graph merges all alternate paths into the reference, hence alternate haplotypes start and end coincident with the reference and global alignment is appropriate.}.
+
+Then, for each row $1 \le i \le {\rm length(reference)}$ we loop over all columns $1 \le j \le {\rm length(alternate)}$ and do the following:
+
+\begin{itemize}
+\item Update deletion scores:  The score for opening a downward gap is $M_{i-1,j} + w_{\rm open}$.  The score for extending an existing deletion is $G_j + w{\rm extend}$, where $D_j$ at this point still pertains to the $previous$ row $i - 1$.  We set $D_j$ (modifying it in-place) to the greater of these values. If the gap-opening score is greater than the gap-extending score, we set $S^d_j = 1$, otherwise we increment $S^d_j$ by 1.
+\item Update insertion scores:  When we begin traversing row $i$ we initialize the current best score for alignments ending in a ``rightward'' gap (i.e. an insertion) as $R = -\infty$ and we initialize the length of the terminal gap in this best-scoring alignment as $S^r = 0$.  Note that these values are local to the inner loop over $j$ and thus do not need to be arrays\footnote{In the code, they \textit{are} arrays, but in the $i$th iteration of the loop over rows, only their $i$th elements are used.  Thus the elements are effectively scalars whose scope is the loop over $j$.}.  At each stage in the loop over $j$ the score for opening a rightward gap is $M_{i,j - 1} + w_{\rm open}$ and score for extending one is $R + w{\rm extend}$, where $R$ still pertains to the previous column $j - 1$.  We set $R$ (modifying it in-place)  to the greater of these values.  If the gap-opening score is greater than the gap-extending score, we set $S^r_j = 1$, otherwise we increment $S^r_j$ by 1.
```

Comment:
This sounds to me like we don't allow multiple insertions or multiple deletions in a single read - is that right? Maybe that's worth a mention

Reply:
We allow multiple indels.  I think the confusion is about the definition of the M_ij matrix.  This is the best score of alignments that are in a match state *at i,j* but which could have indels earlier.  I added a footnote making this explicit.

Reply:
Got it, thanks!